tests/inputFields.spec.ts

```diff
@@ -0,0 +1,105 @@
+import { test, expect } from "@playwright/test";
+
+test.beforeEach(async ({ page }) => {
+  await page.goto("/");
+
+  await page.getByText(" Pet Types").click();
+  const tableTitle = await page.locator("h2").textContent();
+  expect(tableTitle).toEqual("Pet Types");
+});
+
+test("Update pet type", async ({ page }) => {
+  const tableRows = page.locator("tbody tr");
+  await page.waitForSelector('[id="0"]');
+  for (let row of await tableRows.all()) {
+    const cellValue = await row.locator("input").inputValue();
+    if (cellValue === "cat") {
+      await row.getByRole("button", { name: "Edit" }).click();
+      break;
+    }
+  }
+
+  await expect(page.locator("h2")).toHaveText("Edit Pet Type");
+
+  const petTypeInput = page.locator("#name");
+  await petTypeInput.click();
+  await petTypeInput.clear();
+
+  //   await page.waitForTimeout(5000);
+  await petTypeInput.fill("rabbit");
+  await page.getByRole("button", { name: "Update" }).click();
+
+  await page.waitForSelector('[id="0"]');
+  const firstPetTypeInTable = page
+    .locator('input[name="pettype_name"]')
+    .first();
+  await expect(firstPetTypeInTable).toHaveValue("rabbit");
+
+  for (let row of await tableRows.all()) {
+    const cellValue = await row.locator("input").inputValue();
+    if (cellValue === "rabbit") {
+      await row.getByRole("button", { name: "Edit" }).click();
+      break;
+    }
+  }
+  await petTypeInput.click();
+  await petTypeInput.clear();
+  await petTypeInput.fill("cat");
+  await page.getByRole("button", { name: "Update" }).click();
+
+  await expect(firstPetTypeInTable).toHaveValue("cat");
+});
+
+test("Cancel pet type update", async ({ page }) => {
+  const tableRows = page.locator("tbody tr");
+  await page.waitForSelector('[id="0"]');
+  for (let row of await tableRows.all()) {
+    const cellValue = await row.locator("input").inputValue();
+    if (cellValue === "dog") {
+      await row.getByRole("button", { name: "Edit" }).click();
+      break;
+    }
+  }
```

Comment:
I like to see how flexible you are with loops in conditions, and you already have experience with the playwright. With this particular test it could be done a little bit simpler.
Since you have created the locator later in your test for the second table row, you could start your test with it.
`const secondPetTypeInTable = page.locator('tr', {has: page.locator('[id="1"]')});`
Then assert that initial value is "dog"
`await expect(secondPetTypeInTable.locator('input')).toHaveValue("dog")`
Then select the edit button in this row
`await secondPetTypeInTable.getByRole("button", { name: "Edit" }).click()`

This way you simplify the first part of the test and make it more compact. And then later in the test just use the `secondPetTypeInTable` to repeat the validation

Reply:
nice, thank you